ext/pdo_dblib/dblib_driver.c

```diff
@@ -432,12 +434,20 @@ static int pdo_dblib_handle_factory(pdo_dbh_t *dbh, zval *driver_options)
 		}
 	}
 
+	if (!dbh->username && vars[6].optval) {
+		dbh->username = vars[6].optval;
+	}
+
 	if (dbh->username) {
 		if(FAIL == DBSETLUSER(H->login, dbh->username)) {
 			goto cleanup;
 		}
 	}
 
+	if (!dbh->password && vars[7].optval) {
```

Comment:
This might be a dumb question (as I'm new here, still trying to make & run tests properly): how will this behave for `new PDO("...;user=root;password=shouldbeoverwritten", 'overuser', '')`?
As far as I get the key idea here, the 2nd and 3rd construct parameters prevail, so user should be `overuser` and password should be empty. But will the test here consider that since password (from 3rd param) is empty, then the one in DSN will be used and so, instead of "overuser"+"" empty password, we get "overuser"+"shouldbeoverwritten" ?

Reply:
@SpyroTEQ Sorry for the long time between replies, as this PR seems inactive. The idea is that if a username and password is set for the constructor to PDO (argument 2 and 3), then it will override the ones sent in the DSN, this is how other PDO drivers did it so I wrote it like that (in this case for `pdo_dblib`) to keep the drivers somewhat consistent

Reply:
Got (sorry from ym side too to take so long to answer), but I was actually focusing on how you test that the password has been set in DSN and how you test that password has been set as 3rd argument. What I mean is that `!dbh->password` might be `true` if password is not set, but is it also `true` if password is set to empty? Same question for `vars[7].optval`?

In other words, these cases might be worth being tested (or must match the tests used, considering empty/not set values as different):
- `new PDO("...;password=shouldbeoverwritten", 'overuser', '')` : password must be the empty one
- `new PDO("...;password=shouldbeoverwritten", 'overuser', '000')` : password must be '000'
- `new PDO("...;password=shouldbeoverwritten", 'overuser', 'false')` : password must be 'false' (pretty dumb password but still)
- `new PDO("...;password=", 'overuser', 'xxx')` : password must be 'xxx'
- `new PDO("...;password=000", 'overuser', 'xxx')` : password must be 'xxx'
- `new PDO("...;password=inuse", 'overuser')` : password must be 'inuse'
- `new PDO("...;password=", 'overuser')` : password must be empty one
- `new PDO("...;", 'overuser')` : password must be backward compatible

Maybe this is an absurd question in C code, but in PHP code itself, it's often easy to mistake between empty string, not set string (null), 0-filled string, false literal, etc

Reply:
@SpyroTEQ while that is indeed interesting - it is currently impossible to test like that. We have one valid set of credentials for testing and I don't think it's productive to start issuing `SET PASSWORD` calls just for this. Since the implementation is already in C - I'm not sure testing for weird PHP behavior is productive here. I do invite you to test the examples you specified and report back if you find any discrepancies.